### PR TITLE
MAP-3

### PR DESCRIPTION
**Pull request by suniala**

### CODE REVIEW

File: api-common/src/main/java/org/openmrs/module/metadatamapping/MetadataTermMapping.java [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+ */
+public class MetadataTermMapping extends BaseOpenmrsMetadata {
+	
+	private Integer metadataTermMappingId;
+	
+	private MetadataSource metadataSource;
+	
+	private String code;
+	
+	private String metadataClass;
```
Comment (rkorytkowski) [inline api-common/src/main/java/org/openmrs/module/metadatamapping/MetadataTermMapping.java]: In order to efficiently search for mapped items we need to create an index on metadataSource, code and class. The class name may be too long for a DB index (mysql innodb has 255 chars limit per column) so we will need an additional field to store a hashed (e.g. md5) version of metadataClass, which is short enough to be indexed. You can address that as a separate issue.

**Comment by rkorytkowski** [comment]
Merged! Sorry for the delay. I lost it from my radar.

File: api-common/src/main/java/org/openmrs/module/metadatamapping/api/db/hibernate/HibernateMetadataMappingDAO.java [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+		
+		return criteria.list();
+	}
+	
+	@Override
+	public MetadataTermMapping getMetadataTermMapping(MetadataSource metadataSource, String metadataTermCode) {
+		Criteria criteria = sessionFactory.getCurrentSession().createCriteria(MetadataTermMapping.class);
+		criteria.add(Restrictions.eq("metadataSource", metadataSource));
+		criteria.add(Restrictions.eq("code", metadataTermCode));
+		return (MetadataTermMapping) criteria.uniqueResult();
```
Comment (rkorytkowski) [inline api-common/src/main/java/org/openmrs/module/metadatamapping/api/db/hibernate/HibernateMetadataMappingDAO.java]: We'll need DB indexes for the fields that we use in get methods.

File: omod/src/main/resources/liquibase.xml [hunk 3]
```diff
@@ -35,5 +35,151 @@
...
+			</not>
+		</preConditions>
+		<comment>Create metadata source table</comment>
+		<createTable tableName="metadatamapping_metadata_source">
+			<column name="metadata_source_id" autoIncrement="true" type="int">
+				<constraints primaryKey="true" nullable="false" />
+			</column>
+			
+			<!-- columns for OpenMRSMetadata -->
+			<column name="name" type="varchar(255)">
```
Comment (rkorytkowski) [inline omod/src/main/resources/liquibase.xml]: Shouldn't this be unique?

File: api-common/src/main/java/org/openmrs/module/metadatamapping/MetadataTermMapping.java (hunk 1 above)
Comment (suniala) [inline api-common/src/main/java/org/openmrs/module/metadatamapping/MetadataTermMapping.java]: Indexes will be implemented https://issues.openmrs.org/browse/MAP-5
I added your suggestion there as a comment.

File: api-common/src/main/java/org/openmrs/module/metadatamapping/api/db/hibernate/HibernateMetadataMappingDAO.java (hunk 2 above)
Comment (suniala) [inline api-common/src/main/java/org/openmrs/module/metadatamapping/api/db/hibernate/HibernateMetadataMappingDAO.java]: Ditto: https://issues.openmrs.org/browse/MAP-5

File: omod/src/main/resources/liquibase.xml (hunk 3 above)
Comment (suniala) [inline omod/src/main/resources/liquibase.xml]: Yes, it should be as we are using source name as an identifier. I forgot about this when I was implementing the schema as I was concentrating on OpenMRSMetadata semantics. Regarding those semantics, it actually looks like name is sometimes unique (for example in CareSetting) even though it usually is not.

Anyway, I will add a uniqueness constraint on source name.

**Pull request merged**